raiden-ts/src/helpers.ts

```diff
@@ -341,3 +344,51 @@ export const isValidUrl = (url: string): boolean => {
       : /^(?:(http|https):\/\/)?([^\s\/$.?#&"']+\.)*[^\s\/$?#&"']+(?:(\d+))*$/;
   return regex.test(url);
 };
+
+/**
+ * Construct entire ContractsInfo using UserDeposit contract address as entrypoint
+ *
+ * @param provider - Ethers provider to use to fetch contracts data
+ * @param userDeposit - UserDeposit contract address as entrypoint
+ * @returns contracts info, with blockNumber as block of first registered tokenNetwork
+ */
+export async function fetchContractsInfo(
+  provider: JsonRpcProvider,
+  userDeposit: Address,
+): Promise<ContractsInfo> {
+  const userDepositContract = UserDepositFactory.connect(userDeposit, provider);
+
+  const oneToN = (await userDepositContract.one_to_n_address()) as Address;
+  const monitoringService = (await userDepositContract.msc_address()) as Address;
+  const monitoringServiceContract = MonitoringServiceFactory.connect(monitoringService, provider);
+
+  const tokenNetworkRegistry = (await monitoringServiceContract.token_network_registry()) as Address;
+  const tokenNetworkRegistryContract = TokenNetworkRegistryFactory.connect(
+    tokenNetworkRegistry,
+    provider,
+  );
+
+  const secretRegistry = (await tokenNetworkRegistryContract.secret_registry_address()) as Address;
+  const serviceRegistry = (await monitoringServiceContract.service_registry()) as Address;
+
+  const logs = await provider.getLogs({
+    ...tokenNetworkRegistryContract.filters.TokenNetworkCreated(null, null),
+    fromBlock: 1,
+    toBlock: 'latest',
+  });
+  let firstBlock = 0;
+  for (const { blockNumber } of logs) {
+    if (blockNumber) {
+      firstBlock = firstBlock ? Math.min(firstBlock, blockNumber) : blockNumber;
+    }
+  }
```

Comment:
The code looks more complicated than what it actually does. Looks pretty inefficient (code-wise). Couldn't you simply sort the logs by `blockNumber` and take the first/last list entry (depending on order) or `0` as default? This would be about 2 lines of code (before formatting) and much easier to read since it is more declarative (at least for me).

Reply:
I really don't understand how `sort` + `get first` is more inefficient than `Math.min`. I disagree with you on this, although I agree this can be made more explicit with `Math.min` as I did on the other PR. I'll rewrite this then.